File: src/main/java/dev/slne/discord/Bootstrap.java
package dev.slne.discord;

import dev.slne.data.api.DataApi;
import dev.slne.discord.datasource.DiscordDataInstance;
import dev.slne.discord.message.RawMessages;
import java.util.Random;
import lombok.Getter;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import org.springframework.context.ConfigurableApplicationContext;

/**
 * The type Bootstrap.
 */
@Getter
public class Bootstrap {

  private static final ComponentLogger LOGGER = ComponentLogger.logger("Bootstrap");
  @Getter
  private static Random random;

  @Getter
  private static ConfigurableApplicationContext context;
  private final DiscordBot discordBot;

  private DiscordDataInstance dataInstance;

  /**
   * Constructor for the launcher
   */
  public Bootstrap() {
    discordBot = new DiscordBot();

    random = new Random();
  }

  /**
   * Main method
   *
   * @param args The arguments
   */
  public static void main(String[] args) {
    Bootstrap bootstrap = new Bootstrap();

    LOGGER.info("Loading messages...");
    RawMessages.class.getClassLoader();

    bootstrap.onLoad();
    bootstrap.onEnable();
  }

  /**
   * Method called when the launcher is loaded
   */
  public void onLoad() {
    final long start = System.currentTimeMillis();
    dataInstance = new DiscordDataInstance();
    new DataApi(dataInstance);

    context = DiscordSpringApplication.run();

    discordBot.onLoad();
    final long end = System.currentTimeMillis();

    LOGGER.info(
        "Done (%.3fs)! Type 'help' for a list of commands.".formatted((end - start) / 1000.0));
  }

  /**
   * Method called when the launcher is enabled
   */
  public void onEnable() {
    discordBot.onEnable();
  }

  /**
   * Method called when the launcher is disabled
   */
  public void onDisable() {
    discordBot.onDisable();
  }

}

File: src/main/kotlin/dev/slne/discord/Bootstrap.kt
package dev.slne.discord

import com.github.ajalt.clikt.command.parse
import com.github.ajalt.clikt.core.CliktError
import com.github.ajalt.clikt.core.ContextCliktError
import dev.slne.discord.console.buildRootCommand
import dev.slne.discord.discord.interaction.command.DiscordCommandProcessor
import dev.slne.discord.message.RawMessages
import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import kotlinx.coroutines.runBlocking
import net.dv8tion.jda.api.JDA
import net.kyori.adventure.text.logger.slf4j.ComponentLogger
import org.springframework.stereotype.Component
import kotlin.concurrent.thread
import kotlin.system.measureTimeMillis

private val logger = ComponentLogger.logger("Bootstrap")

@Component
class Bootstrap(
    private val jda: JDA,
    private val commandProcessor: DiscordCommandProcessor,
) {

    @PostConstruct
    fun onLoad() {
        val ms = measureTimeMillis {
            Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
                logger.error("Uncaught exception in thread ${thread.name}", throwable)
            }

            logger.info("Loading messages...")
            RawMessages::class.java.getClassLoader()

            listenForCommands(jda, commandProcessor)
        }

        logger.info("Done and ready ({}ms)! Type 'help' for a list of commands.", ms)
    }

    @PreDestroy
    fun onDisable() {
    }
}

private fun listenForCommands(
    jda: JDA,
    commandProcessor: DiscordCommandProcessor,
) = thread {
    val root = buildRootCommand(jda, commandProcessor)

    print("> ")
    while (true) {
        val input = readlnOrNull() ?: continue

        try {
            if (input == "help") {
                root.echoFormattedHelp()
            } else {
                runBlocking { root.parse(input.split(" ")) }
            }
        } catch (error: CliktError) {
            if (error is ContextCliktError) {
                error.context?.command?.echoFormattedHelp(error)
            } else {
                root.echoFormattedHelp(error)
            }

        } catch (exception: Exception) {
            logger.error("Failed to parse command.", exception)
        }

        print("> ")
    }
}


